src/sage/rings/fraction_field_element.pyx

```diff
@@ -446,6 +446,50 @@ cdef class FractionFieldElement(FieldElement):
         """
         return self._numerator(*x, **kwds) / self._denominator(*x, **kwds)
 
+    def subs(self, in_dict=None, *args, **kwds):
+        r"""
+        Substitute variables in the numerator and denominator of ``self``.
+
+        If a dictionary is passed, the keys are mapped to generators
+        of the parent ring.  Otherwise, the arguments are transmitted
+        unchanged to the method ``subs`` of the numerator and the
+        denominator.
+
+        EXAMPLES::
+
+            sage: x, y = PolynomialRing(ZZ, 2, 'xy').gens()
+            sage: f = x^2 + y + x^2*y^2 + 5
+            sage: (1/f).subs(x=5)
+            1/(25*y^2 + y + 30)
+
+        TESTS:
+
+        Check that :issue:`37122` is fixed::
+
+            sage: P = PolynomialRing(QQ, ["x%s" % i for i in range(10000)])
+            sage: PF = P.fraction_field()
+            sage: p = sum(i*P.gen(i) for i in range(5)) / sum(i*P.gen(i) for i in range(8))
+            sage: v = P.gen(4)
+            sage: p.subs({v: 100})
+            (x1 + 2*x2 + 3*x3 + 400)/(x1 + 2*x2 + 3*x3 + 5*x5 + 6*x6 + 7*x7 + 400)
+        """
+        if isinstance(in_dict, dict):
+            gens = self.parent().gens()
+
+            def to_R(m):
+                try:
+                    mi = gens.index(m)
+                except ValueError:
+                    return m
+                return mi
+            in_dict = {to_R(m): v for m, v in in_dict.items()}
```

Comment:
This can be stronger than doing conversion (quite possibly more expensive too) but it could also be independent. The issue is that equality check that is done as part of `index()` goes through coercion. This means that it could potentially construct a pushout and do the comparison in the larger ring. So it might allow for unrelated variables than a strict conversion `R(m)` check would not allow. On the other than, conversion can be much more general depending on the implementation, but I would imagine it covers most reasonable pushout cases that would result in an equality. It is likely to be the most consistent as well.

Also, this implementation might rely on equal elements having equal hashes (which Python says should be true, but coercion and pushouts can lead to messy cases where this unintentionally does not hold).

Because of this, I think you are best simply doing
```python
R = self.parent()
in_dict = {R(m): v for m,. v in in_dict.items()}
```

Reply:
I think we want to restrict keys to generators and indices of generators. I think your suggestion does not allow indices, although that should be easy to fix. However, I would also like to look at a few concrete examples to make up my mind. My experience with the constructors is that they are quite careless in some cases, which is why I prefer to avoid them.

Reply:
Sorry, I was very imprecise and misleading.  What I meant to say is, that we want to restrict keys to things that the user regards as generators.  More to the point:  I think we don't want to allow for things like `{x/y: 1}` in a ring of rational functions.  On the other hand, maybe we should, I don't really know.

The idea to allow integers is to have a possibility of speedup: if the key is an integer, it should be possible to avoid expensive operations completely.  I am well aware of the fact that the current code does not take advantage of this possibility.

It surprises me that conversion might be more consistent than checking equality.  Are you sure about that?

Reply:
To summarize: would you think
```
R = self.parent()
in_dict = {m if isinstance(m, (Integer, int)) else R(m): v for m, v in in_dict.items()}
```
is better?

Reply:
Yes, that is good after a slight variant:
```python
R = self.parent()
in_dict = {ZZ(m) if m in ZZ else R(m): v for m, v in in_dict.items()}
```
to catch things like `{2/2: x+y}`. My previous proposal, in principle, does allow for integers, but it would break over finite fields in a subtle way.

With regards to your other comments, things like `{x/y: 1}` would fail, but perhaps at a later stage when it compares with the generators. For what a user looks like a generator can be very broad. Consider the following example that generates a pushout:
```
sage: R.<x,t> = QQ[]
sage: S.<x> = ZZ['a'][]
sage: x + t
x + t
sage: _.parent()
Multivariate Polynomial Ring in x, t over Univariate Polynomial Ring in a over Rational Field
sage: S.gen() == R.gen(0)
True
```
Is the generator of `S` a generator of `R` and vice versa? The conversion does work in this case. However, what about the string `'x'`?

To put it another way, I think each ring should be responsible for what it considers as input and is the best thing to decide that. Yes, some are imprecise/not-well-coded, but the specifications about what is an element (and hence might be a generator) is more clear.

Reply:
Unfortunately, this does not work:
```
sage -t --warn-long 81.0 --random-seed=222573162246150205167095728432903358574 src/sage/rings/fraction_field_element.pyx
**********************************************************************
File "src/sage/rings/fraction_field_element.pyx", line 374, in sage.rings.fraction_field_element.FractionFieldElement.__hash__
Failed example:
    ((x+1)/(x^2+1)).subs({x:1})
Exception raised:
    Traceback (most recent call last):
      File "/home/martin/sage-trac/src/sage/doctest/forker.py", line 712, in _run
        self.compile_and_execute(example, compiler, test.globs)
      File "/home/martin/sage-trac/src/sage/doctest/forker.py", line 1147, in compile_and_execute
        exec(compiled, globs)
      File "<doctest sage.rings.fraction_field_element.FractionFieldElement.__hash__[15]>", line 1, in <module>
        ((x+Integer(1))/(x**Integer(2)+Integer(1))).subs({x:Integer(1)})
      File "sage/rings/fraction_field_element.pyx", line 481, in sage.rings.fraction_field_element.FractionFieldElement.subs
        num = self._numerator.subs(in_dict, *args, **kwds)
      File "sage/rings/polynomial/multi_polynomial_libsingular.pyx", line 3559, in sage.rings.polynomial.multi_polynomial_libsingular.MPolynomial_libsingular.subs
        raise TypeError("keys do not match self's parent")
    TypeError: keys do not match self's parent
**********************************************************************
File "src/sage/rings/fraction_field_element.pyx", line 474, in sage.rings.fraction_field_element.FractionFieldElement.subs
Failed example:
    p.subs({v: 100})
Exception raised:
    Traceback (most recent call last):
      File "/home/martin/sage-trac/src/sage/doctest/forker.py", line 712, in _run
        self.compile_and_execute(example, compiler, test.globs)
      File "/home/martin/sage-trac/src/sage/doctest/forker.py", line 1147, in compile_and_execute
        exec(compiled, globs)
      File "<doctest sage.rings.fraction_field_element.FractionFieldElement.subs[7]>", line 1, in <module>
        p.subs({v: Integer(100)})
      File "sage/rings/fraction_field_element.pyx", line 481, in sage.rings.fraction_field_element.FractionFieldElement.subs
        num = self._numerator.subs(in_dict, *args, **kwds)
      File "sage/rings/polynomial/multi_polynomial_libsingular.pyx", line 3559, in sage.rings.polynomial.multi_polynomial_libsingular.MPolynomial_libsingular.subs
        raise TypeError("keys do not match self's parent")
    TypeError: keys do not match self's parent
**********************************************************************
2 items had failures:
   1 of  32 in sage.rings.fraction_field_element.FractionFieldElement.__hash__
   1 of   9 in sage.rings.fraction_field_element.FractionFieldElement.subs
    [295 tests, 2 failures, 0.72 s]
----------------------------------------------------------------------
sage -t --warn-long 81.0 --random-seed=222573162246150205167095728432903358574 src/sage/rings/fraction_field_element.pyx  # 2 doctests failed
----------------------------------------------------------------------
```

Reply:
I am guessing that this is because the other instances of `subs` use equality, not conversion.  But I don't know.

Reply:
No, they aren’t unless I am missed something. Sorry, I forgot where this was supposed to live; it should have been `R = self.parent().base()` as it should be a polynomial ring (more generally, the defining domain) generator.